docs/resources/templates/Design-Document-Template.md

```diff
@@ -0,0 +1,52 @@
+
+##	Design
+This section explains the high level changes proposed to achieve the design. Changes should be grouped by project components or in another logical way which makes the design easy to follow.
+If applicable, the design should be broken down into separate phases which each delivers incremental value. The first phase should achieve the minimum requirements outlined in the goals.
+If applicable, alternative approaches should be noted and evaluated for different sections of the design.
+
+##	API Changes
+This is a section detailing any user facing or internal APIs added to components, including configuration changes. Any changes to user facing components are mentioned here (ie. WebUI, command line). If no changes are made, it should be stated in this section.
+Changes should be formatted in a reader friendly manner. A one-cell table should be used to contain the code blocks and monospace font should be used when writing code.
+
+##	Compatibility
+This section explains any compatibility assumptions or limitations. This should apply to compatibility between versions and editions. For designs which affect both enterprise and open source, this section should call out any significant deviations and strategies to avoid divergence as well as the separation between open source and closed source.
+This section also analyzes any dependency changes (ie. pom file changes). In particular, the addition of new external libraries should be justified.
+
+##	Security
+This section explains any security considerations which must be taken into account. If there are no security implications, it should be stated explicitly here.
+
+##	Implementation Detail (Optional)
+This is an OPTIONAL and possibly REPEATED section which dives into a specific implementation detail in the design. Diagrams, code snippets, and other figures are encouraged.
+
+##	Failure Analysis
+This section explains the various failure cases which the design must tolerate as well as failure cases the design neglects to address. Each failure should be noted with the expected behavior as well as ideal behavior, and if different, what effort is required to bridge the gap.
+This section should also outline any potential risks due to the implementation of this design such as performance loss, undesired behavior, and system instability.
+
+##	Test Plan
+This section details the tests which will be added to ensure the quality and maintainability of the implementation. Unit tests should be added under the corresponding Alluxio module. Integration tests should be added to alluxio/tests module. System tests performed manually can be mentioned in the PR. In addition, any new functionality to support testing should be noted.
+
+##	Engineering Estimate
+This section covers the schedule of the design, implementation and testing phases. This will help us estimate the complete time of the feature and align with releases.
+
+##	Sign Off
```

Comment:
I don't know if this section is really necessary if it's something we expect users to upload? I don't really see a straightforward way to modify the document to show we "sign off" on the doc.

If we end up reviewing design docs through PRs I feel it's best to simply use the Review+Approve method on the PR. I actually really like how rust does it. Ex: `https://github.com/rust-lang/rfcs/pull/2867/files`

Reply:
Agreed we should use the Review+Approve method on the PR, which is a placeholder during the feature design period. So the "Sign off" section should be removed from the template.

I like how Rust does this as you recommended. Just curious, how do they discussion on the design MD file?

Reply:
So they have a separate repo where all RFCs get submitted as PRs.

1. Someone submits and RFC containing design of a new feature
2. Reviewers look over the design, add comments (just like in a PR)
2a. Author addresses any comments. If none, go to step 3, otherwise return to 2.
3. Reviewers approve the PR for the RFC, and subsequently merge it.
4. Original author or someone else goes and implements the design outlined in the RFC and submits it to the actual repo

Reply:
Example RFC: `https://github.com/rust-lang/rfcs/pull/2532`

Reply:
Looks really cool! 

I think in a longer-term plan we should have a repository even for people's google docs as design doc. So a design doc linked in the PR at least doesn't get lost anyhow.

Reply:
The issue is that you can't just have a repo for google docs. You need to export them into some format anyways. PDF is the most common. That's why I don't really like the idea of google docs. Especially because the discussion won't occur on github and any history will be hard to track if the doc ever gets deleted or scrubbed by the original author

Reply:
IMO google doc is a good place for ongoing discussions, quick edits (especially most comments can be resolved and not worthy of keeping) and collaborated work. I'm fine with working on shared docs when the PR is still in working phase. In the end when the PR is merged, I agree we should export them. 

On the persistence of the design doc, I think we can just attach the PDF to the PR? 

Overall I think it's hard to have the "full" discussion and find the ideal solution before anyone really follows this procedure. I propose we let it happen first for one or a few PRs and then improve the workflow. Maybe @apc999 what's your opinion on this?